text/0003-remove-handle-ca.md

```diff
@@ -0,0 +1,54 @@
+- Feature Name: handle_ca_alternatives
+- Start Date: 2016-07-10)
+- RFC PR:
+- Issue:
+
+# Summary
+[summary]: #summary
+
+Currently 'Host::Managed' contains `#handle_ca` method. As part of puppet separation effort,
+this method should be moved away from the core.
+The problem is that it is called from a different locations in code: once from `UnattendedController`
+and twice from orchestration - from `ComputeOrchestration` and from `SSHOrchestration`. Those parts
+shouldn't be aware to the fact that puppet is present.
+
+I suggest using [policy-based autosigning](https://docs.puppet.com/puppet/latest/reference/ssl_autosign.html#policy-based-autosigning)
+for making Foreman an authority for rejecting irrelevant nodes.
+
+# Motivation
+[motivation]: #motivation
+
+1. It eliminates the need in `#handle_ca` method, so it would not be available to core or other parts of
+the code.
+2. It improves visibility of rejected autosign requests.
+
+
+# Detailed design
+[design]: #detailed-design
+
+First we will modify puppet CA server to support Foreman as an [external policy provider](https://docs.puppet.com/puppet/latest/reference/ssl_autosign.html#enabling-policy-based-autosigning):
+1. Add a simple shell script that will use curl to call foreman server/proxy and pass the CSR to the request.
+2. Mark the script as a policy executable in the puppet.conf
+
+Next we will add a special controller that will respond to those requests, check for authenticity of the CSR and
+return a result that could be parsed by the script.
+
+\[Optional\]
+We can add information to a node that will be passed as part of the CSR, for example we can use a one-time password that
+will be invalidated upon first CSR validation. This will ensure that each host is registered at most once.
```

Comment:
@timogoebel that's simply the threat of someoe getting a certificate by any means - once they have a cert, they can get a catalog.

I would say the threat model is traffic analysis. There are three paths here:

* A physical (or physical-like VM) will boot, request a template, and continue. This is the weak point - rarely do provisioning systems support HTTPS. Any pre-shared info (or hostnames for traditional autosign) would be visible in the reply.
* A user-data based VM has a similar issue - it requests a finish template. This could be HTTPS, but `unattended_url` isn't that flexible to use HTTP or HTTPS as required.
* An SSH-based VM is probably OK - the VM has a key baked in, and the transfer of templates/data is secure.

So, if the one-time key needs to be present via all 3 methods, we need to figure out how. The first case seems the hardest, but we could potentially embed the key in the PXE template. If we added `foremankey=abcdef` to the APPEND line, it could be queried from `/prc/cmdline` later, right?

This could also be added to the user-data metadata (which is, presumably, going by a secure route to the hypervisor) and to the SSH provision methods. So long as all 3 methods result in the key being in the same place on the host filesystem, further orhestration (puppet, etc) can make use of it.